### PR TITLE
Add OneAgent metadata enrichment

### PR DESCRIPTION
**Pull request by pirgeo**

### CODE REVIEW

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java [hunk 1]
```diff
@@ -44,6 +42,64 @@
...
+   */
+  public static MetricAdapter getInstance() {
+    if (instance == null) {
+      instance = new MetricAdapter();
+    }
+    return instance;
+  }
+
+  /**
+   * contains the static dimensions (e. g. tags and OneAgent metadata enrichment, if set up) that
```
Comment (arminru) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java:61]:
```suggestion
   * Contains the static dimensions (i.e., tags and OneAgent metadata enrichment, if set up) that
```

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java [hunk 2]
```diff
@@ -44,6 +42,64 @@
...
+   * are added to each mint line sent to the ingestion API.
```
Comment (arminru) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java:62]:
```suggestion
   * are added to each MINT line sent to the ingestion API.
```

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java [hunk 3]
```diff
@@ -44,6 +42,64 @@
...
+   */
+  private Collection<Dimension> constantDimensions = null;
+
+  /**
+   * Set the dimensions (as key value pairs) that should be added as dimensions in the mint line.
+   * keys and values are sanitized by {@link #toMintDimension}. Tags specified here are appended
+   * after tags that are set during usage of the instruments that actually do the value recording.
+   * Once the tags are set, they cannot be overwritten and will stay the same for the runtime of the
+   * program.
```
Comment (arminru) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java:71]:
```suggestion
   * Sets the dimensions (as key-value pairs) that should be added as dimensions to all metrics.
   * Keys and values are sanitized by {@link #toMintDimension}. Tags specified here will overwrite
   * any labels added by the application using instruments.
   * These tags can only be set once, so this should be called in the setup.
```

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java [hunk 4]
```diff
@@ -44,6 +42,64 @@
...
+   *
+   * @param tags a list of tags that are appended to each mint line, as key-value pairs (mapping
```
Comment (arminru) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java:73]:
```suggestion
   * @param tags A collection of tags added as dimensions to all metrics, as key-value pairs (mapping
```

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java [hunk 5]
```diff
@@ -44,6 +42,64 @@
...
+      }
+      constantDimensions = Collections.unmodifiableList(localConstantDimensions);
+    } else {
+      logger.warning("overwriting of tags not allowed.");
+    }
+  }
+
+  /**
+   * resets the values stored in the singleton to null and deletes the singleton instance itself.
+   */
```
Comment (arminru) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java:96]:
```suggestion
```
It's package-private anyway and concise enough.

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java [hunk 6]
```diff
@@ -44,6 +42,64 @@
...
+  @VisibleForTesting
+  static void resetForTest() {
+    getInstance().constantDimensions = null;
```
Comment (arminru) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java:99]:
```suggestion
```
Let's have the GC take care of that.

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java [hunk 7]
```diff
@@ -190,6 +246,24 @@ private static void addGaugeData(
     }
   }
 
+  /**
+   * This function first transforms the given labels to dimensions using {@link
+   * #convertLabelsToDimensions(Labels)} and then adds static labels that are stored in the
+   * singleton instance, if there are any.
```
Comment (arminru) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java:252]:
```suggestion
   * singleton instance, if there are any, while overwriting any labels with the same key.
```

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java [hunk 8]
```diff
@@ -190,6 +246,24 @@ private static void addGaugeData(
...
+   *
+   * @param labels the labels to be transformed by {@link #convertLabelsToDimensions}.
+   * @return A list of {@link Dimension} objects, that will be serialized.
```
Comment (arminru) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java:255]:
```suggestion
   * @return A list of {@link Dimension} objects, to be serialized.
```

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java [hunk 9]
```diff
@@ -190,6 +246,24 @@ private static void addGaugeData(
...
+   */
+  private static List<Dimension> getCombinedDimensions(Labels labels) {
+    List<Dimension> dynamicDimensions = convertLabelsToDimensions(labels);
```
Comment (arminru) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java:258]: This should actually be a set. We can't have duplicate dimensions.

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java [hunk 10]
```diff
@@ -190,6 +246,24 @@ private static void addGaugeData(
...
+
+    if (getInstance().constantDimensions != null) {
+      dynamicDimensions.addAll(getInstance().constantDimensions);
```
Comment (arminru) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java:261]: This will add both dimensions instead of overwriting them in case there are duplicate keys.

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java [hunk 11]
```diff
@@ -44,6 +42,64 @@
...
+    if (tags == null) {
+      return;
+    }
+
+    // the constantDimensions field is only populated once, and the dimensions created then are
+    // reused.
+    if (this.constantDimensions == null) {
+      List<Dimension> localConstantDimensions = new ArrayList<>();
+      for (AbstractMap.SimpleEntry<String, String> tag : tags) {
+        localConstantDimensions.add(toMintDimension(tag.getKey(), tag.getValue()));
```
Comment (arminru) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java:86]: Please check if `toMintDimension` might return null or throw for invalid input and handle it appropriately.

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/OneAgentMetadataEnricher.java [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+import java.io.FileReader;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.*;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class OneAgentMetadataEnricher {
```
Comment (arminru) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/OneAgentMetadataEnricher.java:29]:
```suggestion
final class OneAgentMetadataEnricher {
```
No need to expose this. Same for `MetricAdapter`.

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/OneAgentMetadataEnricherTest.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.logging.Logger;
+import org.junit.jupiter.api.Test;
+
+public class OneAgentMetadataEnricherTest {
```
Comment (arminru) [inline dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/OneAgentMetadataEnricherTest.java:13]: Please add tests to verify proper handling of missing/inaccessible/invalid indirection and metadata files.

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java (hunk 10 above)
Comment (arminru) [reply dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java:261]: Please add a test covering that as well :-)

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java [hunk 14]
```diff
@@ -249,42 +248,47 @@ private static void addGaugeData(
...
-        new BiConsumer<String, String>() {
-          @Override
-          public void accept(String k, String v) {
+        (String k, String v) -> {
+          try {
             dimensions.add(toMintDimension(k, v));
+          } catch (DynatraceExporterException dee) {
+            logger.warning(
+                String.format(
+                    "Could not transform '%s/%s' to MINT dimension: %s", k, v, dee.getMessage()));
```
Comment (arminru) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java:286]:
```suggestion
                    "Could not transform OTel label '%s'->'%s' to Dynatrace dimension: %s", k, v, dee.getMessage()));
```
MINT is no customer-facing terminology, it's just the internal abbreviation for the Metrics ingest API.

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java [hunk 15]
```diff
@@ -341,10 +345,15 @@ static String toMintMetricKey(String metricKey) throws DynatraceExporterExceptio
...
    * @return a MINT-compatible Dimension (sanitized key and value).
-   * @throws DynatraceExporterException if an error occured during sanitizing the Strings.
+   * @throws DynatraceExporterException if an error occurred during sanitizing the Strings or if a
+   *     null value was passed.
    */
   static Dimension toMintDimension(String key, String value) throws DynatraceExporterException {
-    return Dimension.create(toMintDimensionKey(key), toMintDimensionValue(value));
+    try {
+      return Dimension.create(toMintDimensionKey(key), toMintDimensionValue(value));
+    } catch (NullPointerException npe) {
```
Comment (arminru) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java:354]: Explicitly null-checking both key and value would be more efficient than having an additional exception thrown.

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/OneAgentMetadataEnricher.java [hunk 16]
```diff
@@ -15,26 +15,25 @@
...
     this.logger = logger;
   }
 
+  Logger getLogger() {
+    return this.logger;
+  }
+
   public Collection<AbstractMap.SimpleEntry<String, String>> getDimensionsFromOneAgentMetadata() {
-    return parseOneAgentMetadata(getMetadataFileContent());
+    String indirectionBaseName = "dt_metadata_e617c525669e072eebe3d0f08212e8f2";
```
Comment (arminru) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/OneAgentMetadataEnricher.java:35]: I'd prefer a class constant.

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/OneAgentMetadataEnricher.java [hunk 17]
```diff
@@ -45,23 +44,24 @@ public OneAgentMetadataEnricher(Logger logger) {
    *
    * @param lines a {@link Collection<String>} containing key-value pairs (as one string) separated
    *     by an equal sign.
-   * @return A {@link Collection<AbstractMap.SimpleEntry<String, String>>}. These represent the the
+   * @return A {@link Collection<AbstractMap.SimpleEntry<String,String>>}. These represent the the
    *     lines passed in separated by the first occurring equal sign on each line, respectively. If
    *     no line is parsable, returns an empty list
    */
+  @SuppressWarnings("JavaDoc")
```
Comment (arminru) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/OneAgentMetadataEnricher.java:51]: What is it complaining about?

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/OneAgentMetadataEnricher.java [hunk 18]
```diff
@@ -70,57 +70,110 @@ public OneAgentMetadataEnricher(Logger logger) {
...
-    String indirectionBaseName = "dt_metadata_e617c525669e072eebe3d0f08212e8f2";
-    String secretFileName = null;
-    try (BufferedReader reader =
-        new BufferedReader(new FileReader(String.format("%s.properties", indirectionBaseName)))) {
-      String line = "";
-      // this whole while block is to make sure this function will still work even if the contents
-      // of the
-      // indirection file were to change in the future.
+  /**
+   * Get the file name of the file to which OneAgent persists its parameters.
```
Comment (arminru) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/OneAgentMetadataEnricher.java:82]:
```suggestion
   * Get the file name of the file in which OneAgent provides the metadata.
```

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/OneAgentMetadataEnricher.java [hunk 19]
```diff
@@ -70,57 +70,110 @@ public OneAgentMetadataEnricher(Logger logger) {
...
+   *
+   * @param fileContents A {@link Reader} object pointing at the indirection file. Will be read
+   *     using a {@link BufferedReader}
```
Comment (arminru) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/OneAgentMetadataEnricher.java:85]:
```suggestion
   * @param fileContents A {@link Reader} object pointing at the indirection file.
```
Implementation detail

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/OneAgentMetadataEnricher.java [hunk 20]
```diff
@@ -70,57 +70,110 @@ public OneAgentMetadataEnricher(Logger logger) {
...
+    String line;
+
+    try (BufferedReader reader = new BufferedReader(fileContents)) {
       // read file line by line, and stop if the end of the file is reached.
       while ((line = reader.readLine()) != null) {
+        line = line.trim();
         // the secret file contains the basename and a random number at the end.
-        if (line.contains(indirectionBaseName)) {
-          secretFileName = line;
+        if (line.startsWith(prefix)) {
```
Comment (arminru) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/OneAgentMetadataEnricher.java:110]: I don't think you can rely on the actual file name to start with a particular prefix, that's not defined in the OneAgent API and therefore subject to change.

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/NormalizationTest.java [hunk 21]
```diff
@@ -1,4 +1,4 @@
-/**
+/*
```
Comment (arminru) [inline dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/NormalizationTest.java:1]:
```suggestion
/**
```

File: dynatrace/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker [hunk 22]
```diff
@@ -0,0 +1 @@
+mock-maker-inline
```
Comment (arminru) [inline dynatrace/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:1]: Is this necessary?

Comment (pirgeo) [reply dynatrace/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:1]: yes, this tells `mockito` to allow mocking of `final` classes and has to be enabled like this since it is an experimental feature.

**Comment by pirgeo** [comment]
In `MetricAdapter.java`, lines 132-142 in the latest version, the metric line is generated once for each metric. That means that if one Dimension parsing failure occurs, the Exception will bubble up to this `try-catch` and the whole metric line will be discarded, even if other dimensions are valid. There doesn't seem to be a standardized behavior for that (in python for example invalid leading characters are simply trimmed and the other dimensions are added regardless).

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java [hunk 23]
```diff
@@ -249,19 +250,20 @@ private static void addGaugeData(
    * This function first transforms the given labels to dimensions using {@link
    * #convertLabelsToDimensions(Labels)} and then adds static labels that are stored in the
    * singleton instance, if there are any. If two labels have the same key, the last added label is
-   * retained. Constant dimensions such as tags and OneAgent metadata labels are added last.
+   * retained. Constant dimensions such as tags and OneAgent metadata labels are added last and will
+   * overwrite existing labels.
```
Comment (arminru) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java:254]:
```suggestion
   * overwrite the labels passed to this method.
```

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java [hunk 24]
```diff
@@ -249,19 +250,20 @@ private static void addGaugeData(
...
    *
    * @param labels the labels to be transformed by {@link #convertLabelsToDimensions}.
    * @return A list of {@link Dimension} objects to be serialized.
```
Comment (arminru) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java:257]:
```suggestion
   * @return An unmodifiable list of {@link Dimension} objects to be serialized.
```

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/OneAgentMetadataEnricher.java [hunk 25]
```diff
@@ -22,6 +21,7 @@
 
 final class OneAgentMetadataEnricher {
   private final Logger logger;
+  private static final String indirectionBaseName = "dt_metadata_e617c525669e072eebe3d0f08212e8f2";
```
Comment (arminru) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/OneAgentMetadataEnricher.java:24]:
```suggestion
  private static final String INDIRECTION_FILE_NAME = "dt_metadata_e617c525669e072eebe3d0f08212e8f2";
```

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/OneAgentMetadataEnricher.java [hunk 26]
```diff
@@ -79,25 +79,16 @@ Logger getLogger() {
...
-   * @param indirectionBaseName The prefix for the OneAgent metadata file. Lines in the indirection
-   *     file without this prefix are ignored. If null is passed as this parameter,
-   *     'dt_metadata_e617c525669e072eebe3d0f08212e8f2' is used.
+   * @param fileContents A {@link Reader} object pointing at the indirection file.
    * @return The string containing the filename, with no leading or trailing whitespace.
    * @throws IOException if an error occurs during reading of the file.
    */
-  String getIndirectionFilename(Reader fileContents, String indirectionBaseName)
-      throws IOException {
+  String getIndirectionFilename(Reader fileContents) throws IOException {
```
Comment (arminru) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/OneAgentMetadataEnricher.java:88]: This returns the file name of the metadata file itself, not of the static pre-defined indirection file name (which is already stored in the constant above).

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/OneAgentMetadataEnricher.java [hunk 27]
```diff
@@ -106,8 +97,7 @@ String getIndirectionFilename(Reader fileContents, String indirectionBaseName)
       // read file line by line, and stop if the end of the file is reached.
       while ((line = reader.readLine()) != null) {
```
Comment (arminru) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/OneAgentMetadataEnricher.java:98]: There should only be one file name in there anyway with no leading empty lines so I'd only check the first line and null if there's none.

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/OneAgentMetadataEnricher.java [hunk 28]
```diff
@@ -79,25 +79,16 @@ Logger getLogger() {
...
-   * Get the file name of the file to which OneAgent persists its parameters.
+   * Get the file name of the file in which OneAgent provides the metadata.
    *
-   * @param fileContents A {@link Reader} object pointing at the indirection file. Will be read
-   *     using a {@link BufferedReader}
-   * @param indirectionBaseName The prefix for the OneAgent metadata file. Lines in the indirection
-   *     file without this prefix are ignored. If null is passed as this parameter,
-   *     'dt_metadata_e617c525669e072eebe3d0f08212e8f2' is used.
+   * @param fileContents A {@link Reader} object pointing at the indirection file.
    * @return The string containing the filename, with no leading or trailing whitespace.
```
Comment (arminru) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/OneAgentMetadataEnricher.java:85]:
```suggestion
   * @return The string containing the filename or null if the file is empty.
```
You don't actually trim the string and even if you would it should be needless to say.

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/MetricAdapterTest.java [hunk 29]
```diff
@@ -168,4 +168,113 @@ public void TestAddOneAgentDimensionsEmptyList() {
...
+    MetricAdapter.getInstance().setTags(tags);
+
+    List<Dimension> expected =
+        Arrays.asList(
+            Dimension.create("dim1", "dv1"),
+            Dimension.create("dim2", "dv2"),
+            Dimension.create("tag1", "tv1"),
+            Dimension.create("tag2", "tv2"));
+    // the array list wrap is so the list is sortable as it is otherwise unmodifiable.
+    List<Dimension> got = new ArrayList<>(MetricAdapter.getUniqueCombinedDimensions(labels));
```
Comment (arminru) [inline dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/MetricAdapterTest.java:214]:
```suggestion
    List<Dimension> actual = new ArrayList<>(MetricAdapter.getUniqueCombinedDimensions(labels));
```
is commonly used for that

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java [hunk 30]
```diff
@@ -77,18 +77,19 @@ public void setTags(Collection<AbstractMap.SimpleEntry<String, String>> tags) {
 
     // the constantDimensions field is only populated once, and the dimensions are reused.
     if (this.constantDimensions == null) {
```
Comment (arminru) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/MetricAdapter.java:79]: Please add tests for `setTags(null)` and also for calling `setTags` twice to ensure that doesn't throw.

File: dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/NormalizationTest.java [hunk 31]
```diff
@@ -105,10 +104,8 @@ public void dimensionTest() throws DynatraceExporterException {
   @Test
   public void dimensionKeyTest() throws DynatraceExporterException {
     assertThrows(DynatraceExporterException.class, () -> MetricAdapter.toMintDimension(".", "b"));
-    MetricAdapter.toMintDimensionKey("test..e12");
     assertThrows(DynatraceExporterException.class, () -> MetricAdapter.toMintDimension(".a", "b"));
     assertThrows(DynatraceExporterException.class, () -> MetricAdapter.toMintDimension("a.", "b"));
-    MetricAdapter.toMintDimension("a..b", "b");
```
Comment (arminru) [inline dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/NormalizationTest.java:111]: Why were these removed? I think they're supposed to test that these _don't_ throw. Verifying the return value would be even better, of course.

Comment (pirgeo) [reply dynatrace/src/test/java/com/dynatrace/opentelemetry/metric/NormalizationTest.java:111]: the update to normalization for the metrics in java is still outstanding. Some normalization is already done, but i am not sure if the normalization step follows all the latest guidelines for sanitization. I added the tests back, for now, but we will have to address the normalization step at some point in the future, possibly when adding tags and prefix options.

File: example/src/main/java/com/dynatrace/opentelemetry/metric/example/DynatraceExporterExample.java [hunk 32]
```diff
@@ -55,20 +59,22 @@ public static void main(String[] args) throws Exception {
     // Create a counter
     LongCounter counter =
         meter
-            .longCounterBuilder("example_counter")
+            .longCounterBuilder("otel.dynatrace.java.example_counter")
```
Comment (arminru) [inline example/src/main/java/com/dynatrace/opentelemetry/metric/example/DynatraceExporterExample.java:62]: It's `otel.python.*` in our Python exporter

```suggestion
            .longCounterBuilder("otel.java.example_counter")
```

File: dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporter.java [hunk 33]
```diff
@@ -46,9 +59,11 @@ public static Builder builder() {
   public static DynatraceMetricExporter getDefault() {
     Builder builder = new Builder();
     try {
-      builder.setUrl(new URL("http://127.0.0.1:14499/metrics/ingest")).build();
+      builder
+          .setUrl(new URL("http://127.0.0.1:14499/metrics/ingest"))
+          .setEnrichWithOneAgentMetaData(true);
     } catch (Exception e) {
```
Comment (arminru) [inline dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporter.java:65]:
```suggestion
    } catch (MalformedURLException e) {
```
Should be sufficient, right?

**Review by arminru — APPROVED** [review]
LGTM! 👍

**Comment by pirgeo** [comment]
:shipit: :rocket:
